storage/innobase/lock/lock0lock.cc

```diff
@@ -1009,8 +1012,11 @@ static void lock_wait_wsrep(trx_t *trx)
         lock= lock_rec_get_next(heap_no, lock);
       do
       {
-        /* if victim has also BF status, but has earlier seqno, we have to wait */
-        if (lock->trx != trx &&
+        /* Victim trx needs to be different from BF trx and it has to have a
+           THD so that we can kill it. If victim has also BF status,
+           but has earlier seqno, we have to wait.
+        */
+        if (lock->trx != trx && lock->trx->mysql_thd &&
```

Comment:
The commit message fails to explain what the exact scenario is. For the case `!lock->trx->mysql_thd`, I think that we need a debug assertion to document the assumptions, for example, that this is a recovered transaction, or a transaction in `XA PREPARE` state.

Reply:
Based on some additional detail in [MDEV-33278](https://jira.mariadb.org/browse/MDEV-33278), the scenario is that `lock->trx` is a recovered transaction that has not yet been rolled back. Unless it is in `XA PREPARE` state, it would eventually be rolled back and the lock conflict would be resolved. If it is in `XA PREPARE` state, then the conflict would remain blocking this transaction `trx` until an `XA COMMIT` or `XA ROLLBACK` statement is executed in some other connection.

I see that `trx_t::commit_in_memory()` is clearing both `trx_t::state` and `trx_t::is_recovered` before it invokes `lock_release(trx_t*)` (which would be blocked by the exclusive `lock_sys.latch` that we are holding here). Hence, it is not possible to write a debug assertion to document this scenario.

Please clarify the comment and the commit message about this.